### PR TITLE
Add support for Streams

### PR DESCRIPTION
**Pull request by CoenWarmer**
Resolves https://github.com/elastic/streams-program/issues/393

## Summary

Adds a new `elasticstack_kibana_stream` resource for managing Kibana [Streams](https://www.elastic.co/docs/api/doc/kibana/group/endpoint-streams) — an experimental Kibana feature for managing data ingestion pipelines. Requires Elastic Stack **9.2.0 or higher**.

Three stream types are supported by a single resource:

| Type | Create | Update | Delete | Notes |
|---|---|---|---|---|
| **Wired** | ✅ | ✅ | ✅ | Fully managed data streams with typed field mappings and routing rules |
| **Query** | ✅ | ✅ | ✅ | Virtual streams defined by an ES\|QL query |
| **Classic** | ❌ | ✅ | ❌ | Pre-existing ES data streams, managed via `terraform import` |

### New resource: `elasticstack_kibana_stream`

- **`description`** attribute: optional string description for the stream; defaults to empty string if omitted.
- **`wired_config`** block: manages field type mappings (`fields_json`), routing rules (`routing_json`), processing pipeline steps (`processing_steps_json`), lifecycle policy (`lifecycle_json`), failure store (`failure_store_json`), and index settings (`index_number_of_shards`, `index_number_of_replicas`, `index_refresh_interval`).
- **`classic_config`** block: manages processing steps and field overrides (`field_overrides_json`) for pre-existing Elasticsearch data streams. Attempting to `create` (rather than `import`) a classic stream returns a clear error with import instructions.
- **`query_config`** block: manages the ES\|QL query (`esql`) and optional view name (`view`) for virtual query streams.
- **`queries`** attribute: attaches ES\|QL queries (with `id`, `title`, `description`, `esql`, optional `severity_score` and `evidence`) to any stream type.
- **`dashboards`** attribute: links dashboard IDs to a stream.
- Composite `space_id/name` ID, consistent with other Kibana resources.
- `terraform import` support for all stream types.
- Min version (`9.2.0`) enforced at `Create` time via `EnforceMinVersion`.

### New API client wrapper

`internal/clients/kibanaoapi/streams.go` — thin wrapper over the generated `kbapi` client exposing `GetStream`, `UpsertStream`, and `DeleteStream`.

> **Note on generated client**: `UpsertStream` uses `PutStreamsNameWithBodyWithResponse` (raw bytes) rather than the typed `PutStreamsNameWithResponse` path. The generated `kbapi.PutStreamsNameJSONRequestBody` is a `anyOf` union wrapper with an unexported `union json.RawMessage` field; `encoding/json` cannot marshal unexported fields, so the typed path would send an empty body. This is a known `oapi-codegen` limitation tracked in [oapi-codegen#1665](https://github.com/oapi-codegen/oapi-codegen/issues/1665), targeted for v2.7.0. Once that ships and the kbapi client is regenerated, the raw-bytes workaround can be replaced with the typed path.
>
> The upstream Kibana PR [elastic/kibana#256682](https://github.com/elastic/kibana/pull/256682) (which added `type` discriminator fields to stream definitions) has since merged and the kbapi has been regenerated with the updated spec. The discriminator types are now present in the generated client, but the union wrapper marshaling issue above means the workaround is still required.

> [!NOTE]
> ### Add `elasticstack_kibana_stream` resource for managing Kibana Streams
> - Introduces a new experimental Terraform resource for Kibana Streams, supporting wired, classic, and query stream types via `wired_config`, `classic_config`, and `query_config` blocks.
> - Adds CRUD operations backed by new `GetStream`, `UpsertStream`, and `DeleteStream` client functions in [streams.go](https://github.com/elastic/terraform-provider-elasticstack/pull/1864/files#diff-8cc348bf92347551dfcec04ec5a2f9887bf48bdb6763613b4ff2e6180a49d940); the resource ID is a composite `space_id/name`.
> - Classic streams cannot be created via Terraform (only imported); attempting to create one returns a diagnostic error. Deleting a classic stream only removes it from state without calling the API.
> - Supports attaching ES|QL queries and dashboards to streams, and enforces a minimum Kibana server version of 9.2.0.
> - Risk: The resource is registered as experimental; the schema uses raw JSON strings for several fields (`processing_steps_json`, `routing_json`, etc.), which bypasses plan-time validation of those values.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized afac2b4.</sup>
> 
>

### CODE REVIEW

File: internal/kibana/streams/models_classic.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	}
+
+	// Index settings
+	if ingest.Settings.IndexNumberOfShards != nil {
+		if v, ok := ingest.Settings.IndexNumberOfShards.Value.(float64); ok {
+			m.IndexNumberOfShards = types.Int64Value(int64(v))
+		}
+	} else {
+		m.IndexNumberOfShards = types.Int64Null()
+	}
```
[the remark this answers is not in the document]
Comment (CoenWarmer) [reply internal/kibana/streams/models_classic.go:85]: Addressed with ea32ac0

File: internal/clients/kibanaoapi/streams.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+type StreamIngest struct {
+	Processing   StreamProcessing     `json:"processing"`
+	Settings     StreamIngestSettings `json:"settings,omitempty"` //nolint:modernize
+	Lifecycle    json.RawMessage      `json:"lifecycle,omitempty"`
+	FailureStore json.RawMessage      `json:"failure_store,omitempty"`
+	// Wired is populated only for wired streams.
+	Wired *StreamIngestWired `json:"wired,omitempty"`
+	// Classic is populated only for classic streams.
+	Classic *StreamIngestClassic `json:"classic,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (CoenWarmer) [reply internal/clients/kibanaoapi/streams.go:74]: I partially agree, but the impact is limited. Processing is non-pointer so it always appears in the JSON, but since the API spec lists processing.steps as required, the API likely expects processing to be present on every write. Sending "processing": {} when no steps are configured is correct — it means "clear processing steps".

File: internal/kibana/streams/models_classic.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// toAPIIngest converts the classic config model to an API ingest object.
+func (m *classicConfigModel) toAPIIngest(_ *diag.Diagnostics) *kibanaoapi.StreamIngest {
+	ingest := &kibanaoapi.StreamIngest{}
+
+	// Processing steps
+	if typeutils.IsKnown(m.ProcessingStepsJSON) {
+		ingest.Processing.Steps = json.RawMessage(m.ProcessingStepsJSON.ValueString())
+	}
```
[the remark this answers is not in the document]
Comment (CoenWarmer) [reply internal/kibana/streams/models_classic.go:116]: For classic streams the ingest block is always expected by the API. But it's still cleaner to return nil when truly nothing is configured.

File: internal/kibana/streams/models_wired.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+		ingest.Wired.Fields = json.RawMessage(m.FieldsJSON.ValueString())
+	}
+	if typeutils.IsKnown(m.RoutingJSON) {
+		var routing []kibanaoapi.StreamRoutingRule
+		if err := json.Unmarshal([]byte(m.RoutingJSON.ValueString()), &routing); err != nil {
+			diags.AddError("Failed to parse routing_json", err.Error())
+			return ingest
+		}
+		ingest.Wired.Routing = routing
+	}
```
[the remark this answers is not in the document]
Comment (CoenWarmer) [reply internal/kibana/streams/models_wired.go]: Addressed with 853e75023141e2ae78ac3185a619f17b232f6ea5

File: internal/kibana/streams/schema.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+					"Mutually exclusive with `wired_config` and `classic_config`.",
+				Optional:   true,
+				Attributes: getQueryConfigSchema(),
+				Validators: []validator.Object{
+					objectvalidator.ConflictsWith(
+						path.MatchRoot("wired_config"),
+						path.MatchRoot("classic_config"),
+					),
+				},
+			},
```
[the remark this answers is not in the document]
Comment (CoenWarmer) [reply internal/kibana/streams/schema.go:114]: Addressed with 853e75023141e2ae78ac3185a619f17b232f6ea5

File: internal/kibana/streams/create.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	name := planModel.Name.ValueString()
+
+	apiReq := planModel.toAPIUpsertRequest(ctx, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	_, upsertDiags := kibanaoapi.UpsertStream(ctx, kibanaClient, spaceID, name, apiReq)
+	resp.Diagnostics.Append(upsertDiags...)
+	if resp.Diagnostics.HasError() {
```
[the remark this answers is not in the document]
Comment (CoenWarmer) [reply internal/kibana/streams/create.go:83]: Addressed with 853e75023141e2ae78ac3185a619f17b232f6ea5

File: internal/kibana/streams/schema.go [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+
+// getWiredConfigSchema returns the schema for wired stream configuration.
+func getWiredConfigSchema() map[string]schema.Attribute {
+	return map[string]schema.Attribute{
+		"processing_steps_json": schema.StringAttribute{
+			MarkdownDescription: "Processing pipeline steps as a JSON array. Each step defines a processing rule " +
+				"(e.g. Grok, dissect, rename). Steps are applied in order during ingest.",
+			CustomType: jsontypes.NormalizedType{},
+			Optional:   true,
+		},
```
Comment (tobio) [inline internal/kibana/streams/schema.go:137]: IIUC this would look something like 

```hcl
processing_steps_json = jsonencode([{...}, {...}])
```

I wonder if it would be a nicer UX with minimal effort on our side to use a TF list of JSON strings here instead. It would likely give users much nicer diffs during planning, breaking things down to an individual processor. 

```suggestion
		"processing_steps": schema.ListNestedAttribute{
			MarkdownDescription: "Processing pipeline steps as a JSON array. Each step defines a processing rule " +
				"(e.g. Grok, dissect, rename). Steps are applied in order during ingest.",
			Optional: true,
			NestedObject: schema.NestedAttributeObject{
				Attributes: map[string]schema.Attribute{
					"json": schema.StringAttribute{
						MarkdownDescription: "Processing pipeline step as a JSON object.",
						CustomType:          jsontypes.NormalizedType{},
						Required:            true,
					},
				},
			},
		},
```

File: internal/kibana/streams/models_query.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+func (m *queryConfigModel) populateFromAPI(q *kibanaoapi.StreamQueryESQLDef) {
+	if q == nil {
+		return
+	}
+	m.Esql = types.StringValue(q.Esql)
+	if q.View != "" {
+		m.View = types.StringValue(q.View)
+	} else {
+		m.View = types.StringNull()
+	}
```
Comment (tobio) [inline internal/kibana/streams/models_query.go:41]: Terraform generally expects the final state to equal the planned value, and errors if it doesn't. 

I expect we'll hit an error here if the config has `view = ""` since it will materialise as null in state.

File: internal/kibana/streams/update.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+package streams
+
+import (
+	"context"
+
+	kibanaoapi "github.com/elastic/terraform-provider-elasticstack/internal/clients/kibanaoapi"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+)
+
+func (r *Resource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
```
Comment (tobio) [inline internal/kibana/streams/update.go:27]: This is at least very similar to `Create`, can we have an `upsert` helper here so `Create` and `Update` share the same codepath?

**Review by tobio — COMMENTED** [review]
> Note on generated client: UpsertStream uses PutStreamsNameWithBodyWithResponse (raw bytes) rather than the typed PutStreamsNameWithResponse path. The generated kbapi.PutStreamsNameJSONRequestBody is a anyOf union wrapper with an unexported union json.RawMessage field; encoding/json cannot marshal unexported fields, so the typed path would send an empty body. This is a known oapi-codegen limitation tracked in https://github.com/oapi-codegen/oapi-codegen/issues/1665, targeted for v2.7.0. Once that ships and the kbapi client is regenerated, the raw-bytes workaround can be replaced with the typed path.
> 
> The upstream Kibana PR https://github.com/elastic/kibana/pull/256682 (which added type discriminator fields to stream definitions) has since merged and the kbapi has been regenerated with the updated spec. The discriminator types are now present in the generated client, but the union wrapper marshaling issue above means the workaround is still required.

From the Kibana spec:

```
/api/streams/{name}:
  put:
    ...
      requestBody:
        content:
          application/json:
            schema:
              anyOf:
                - anyOf:
                    - allOf:
...
```

This is pretty horrible, and there's a bunch of other things we could improve here (i.e hinting `number_of_shards` etc are integers). Considering this is actively being developed, can we spend some time on not providing users with an essentially unusable API spec.

This PR needs a whole bunch of acceptance tests, we should have at a minimum:
* One per stream type
* At least one test covering the import path

[schema-coverage](https://github.com/elastic/terraform-provider-elasticstack/tree/main/.agents/skills/schema-coverage) does a pretty good job of suggesting what's missing.

File: internal/kibana/streams/update.go (hunk 9 above)
Comment (CoenWarmer) [reply internal/kibana/streams/update.go:27]: Good idea

File: internal/kibana/streams/models_query.go (hunk 8 above)
Comment (CoenWarmer) [reply internal/kibana/streams/models_query.go:41]: Fixed

File: internal/kibana/streams/schema.go (hunk 7 above)
Comment (CoenWarmer) [reply internal/kibana/streams/schema.go:137]: Good idea
